### PR TITLE
Plugins: Stick header to the top when scrolling down

### PR DESCRIPTION
**Pull request by xavier-lc**
As discussed on https://github.com/Automattic/dotcom-forge/issues/7491#issuecomment-2155360441, we want the header on the `/plugins` page to be sticky, as it already works on the logged out page.

## Proposed Changes

* Add a `useIsVisible` hook that uses an `IntersectionObserver` to determine whether the header needs to be stuck to the top of the page.

## Testing Instructions

* Apply `Hide whitespace` to make the code diff look better.

* Open the live preview
* Go to `/plugins`, check that scrolling down sticks the search and categories to the top, and that scrolling back undoes that:

plugins

* Go to `/plugins/{siteId}`, check that scrolling down sticks the search and categories to the top, and that scrolling back undoes that:

site view

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

File: client/my-sites/plugins/plugins-browser/index.jsx [hunk 1]
```diff
@@ -163,14 +167,12 @@ const PluginsBrowser = ( { trackPageViews = true, category, search, hideHeader }
 				}
 			/>
 
-			{ ! hideHeader && (
```
Comment (xavier-lc) [inline client/my-sites/plugins/plugins-browser/index.jsx:166]: I removed this prop after checking that it wasn't used anywhere on the codebase.

File: client/my-sites/plugins/search-categories/index.tsx [hunk 2]
```diff
@@ -120,45 +111,42 @@ const SearchCategories: FC< {
...
-							width={ width }
-						/>
-					</>
-				) : (
-					<div ref={ categoriesRef }>
-						<Categories selected={ category } noSelection={ searchTerm ? true : false } />
-					</div>
-				) }
-			</div>
-			<div className="search-categories__sticky-ref" ref={ stickySearchBoxRef } />
```
Comment (xavier-lc) [inline client/my-sites/plugins/search-categories/index.tsx:160]: Element extracted to `PluginsBrowser`, since the sticky header logic is managed there.

**Comment by rcrdortiz** [comment]
When scrolling, the header gets completely out of view before popping in.

https://github.com/Automattic/wp-calypso/assets/1989914/01ec08bb-29e8-4b30-a873-ab7b95e72894

**Comment by rcrdortiz** [comment]
At certain widths, we can only fix one element in the menu. Would it make sense to use the mobile query sooner, what do you think about it?
Screenshot 2024-06-19 at 18 05 31

**Comment by xavier-lc** [comment]
> When scrolling, the header gets completely out of view before popping in. I was expecting a behavior similar to CSS sticky instead.

Good point, I've added maring on https://github.com/Automattic/wp-calypso/pull/91946/commits/46f1b2e74fb3d8abc364019922e14b78893de2be so the sticking happens before the search disappears

**Comment by xavier-lc** [comment]
> At certain widths, we can only fix one element in the menu.

What do you mean with "we can only fix one element in the menu."?

**Comment by rcrdortiz** [comment]
> > At certain widths, we can only fix one element in the menu.
> 
> What do you mean with "we can only fix one element in the menu."?

It's a typo sorry, fit instead of fix.

**Comment by rcrdortiz** [comment]
Great work! It's almost perfect now! When scrolling up the menu pops out.


https://github.com/Automattic/wp-calypso/assets/1989914/5ca51fc7-1fbc-4841-8386-e0d65f0f9033

**Comment by xavier-lc** [comment]
> It's a typo sorry, fit instead of fix.

Ah, ok. I think it depends on how long are the category names, though. If they are single word up to 4 can fit. Also, anything below the biggest ipad will already get the dropdown instead of the slider, so I'd leave it as is.

image

> When scrolling up the menu pops out.

Ah, true! I've added a whitespace placeholder on https://github.com/Automattic/wp-calypso/pull/91946/commits/6d3749d99556ce7f8f3a6a0371657a1b1abdb055 that covers the reduced size of the search/categories when they are fixed to the top.

whitespace

**Review by rcrdortiz — APPROVED** [review]
Tests well, thanks for all the work you've done on this!

**Pull request merged**